src/librustc_mir/interpret/cast.rs

```diff
@@ -134,7 +134,7 @@ impl<'mir, 'tcx, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
             layout::Variants::Single { index } => {
                 if let Some(discr) = src.layout.ty.discriminant_for_variant(*self.tcx, index) {
                     assert!(src.layout.is_zst());
-                    return Ok(Scalar::from_uint(discr.val, dest_layout.size).into());
+                    return Ok(self.cast_from_int(discr.val, src.layout, dest_layout)?.into());
```

Comment:
Hm this intuitively seems reasonable... @eddyb is this (more) correct?

Reply:
Probably, but instead of `src.layout` this should use `self.layout_of(discr.ty)` or similar.

Judging by the `cast_from_int` code, the PR as-is would still produce the wrong value for casting to `i128`.

[And I just checked and it's broken for runtime code too](https://play.rust-lang.org/?version=nightly&mode=release&edition=2018&gist=2cfe79f47d7b241d9e0e745f7905ca0e).

Reply:
For context, [this is with both univariant and regular `enum`](https://play.rust-lang.org/?version=nightly&mode=release&edition=2018&gist=9eb37ff88a11ae7a0b48c376ea0265e4).

Reply:
So the discriminant has a different layout that the enum, even though the enum consists only of the discriminant? Nasty.

> And I just checked and it's broken for runtime code too

oO

Reply:
>  this is with both univariant and regular enum.

So with regular enums it is working correctly, is what you are saying? ("This is" sounds like you are saying both are bugged.)

Reply:
Maybe `ty::layout::Abi` having an `is_signed` method was a bad idea, `cast_from_int` should IMO ICE if `src_layout.abi` isn't `Abi::Scalar`.

Reply:
> cast_from_int should IMO ICE if src_layout.abi isn't Abi::Scalar

That sounds reasonable, yes.

> Maybe ty::layout::Abi having an is_signed method was a bad idea

We could make that method ICE if the abi isn't `Scalar`?

Reply:
@RalfJung Sorry, "this is" just referred to the link, a non-univariant enum works fine as you can see if you run the test.

> So the discriminant has a different layout that the enum, even though the enum consists only of the discriminant? Nasty.

Univariant enums have no discriminant representation, be it tag or otherwise, and `discr` is the discriminant value of the type that it's type-checked as (`isize` by default, changes with `#[repr(iN)]`, but I believe `#[repr]` also disables ZST univariant `enum`s)

Reply:
> We could make that method ICE if the `abi` isn't `Scalar`?

I just checked and only miri appears to use that method. Consider checking if the type is signed instead, I guess.

Reply:
> Consider checking if the type is signed instead, I guess.

What do you mean? Using [this](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/ty/struct.TyS.html#method.is_signed)? Note that this also does not complain about use on non-integer types.

Reply:
@eddyb do you think we can make `TyS::is_signed` ICE on non-integer? It seems to have the exact same problem (and just switching to it here would not have prevented this bug).

Reply:
Fair, `cast_from_int` should exhaustively match on `Ty::Int(_)` and `ty::UInt(_)` and not allow any other input types.

Reply:
And you are sure all the other users of that method are correct?^^

Reply:
> And you are sure all the other users of that method are correct?^^

I haven't checked, no, an audit should probably be done.

Reply:
I tried to add the cases mentioned above to the test ([here](https://github.com/rust-lang/rust/pull/70126/commits/14e7cf6d4e2699f1736f86e2c62084cf68d255d8)) but the test still passes. Am I missing a failing case or does this already handle them all?

Reply:
So maybe the run-time version of this is entirely unreachable now, and @eddyb was acutally still testing the Miri version above? As @eddyb said, the input here is a ZST so this can be const-prop'ed even if the input is "unknown".

Reply:
However, I am getting the wrong behavior even with `-Zmir-opt-level=0`, and in that case there should be no const-prop...

Reply:
> As @eddyb said, the input here is a ZST so this can be const-prop'ed even if the input is "unknown".

It also means codegen could have the exact same bug.

Reply:
But if it did, then why would this test pass now after @wesleywiser only fixed Miri?

Reply:
Uh oh, I missed the fact that #70114 affects stable, sorry.

Reply:
Sorry, I just realized I was running an older version of the test locally which is why it kept passing. With that corrected, the casts to `i128` are failing.

Reply:
The commit I just pushed fixes both miri and codegen and the new test passes on all cases.

I believe the conclusion from the rest of this thread was that `cast_from_int` should check the `src_layout.ty` and only allow `ty::Int(_) | ty::UInt(_)`. I tried that but the compiler ICEs during bootstrap unless I also allow `ty::Char | ty::Bool`. Even then, various tests fail because they call `cast_from_int` with sources like enums with explicit discriminants and pointer types. 

Is adding more validation to `cast_from_int` still the right thing to do?

Reply:
> various tests fail because they call `cast_from_int` with sources like enums with explicit discriminants and pointer types.

Both of those examples sound wrong, the `tag` field of an `enum` should have an appropriate type (from `#[repr]` or `isize` by default), and pointers are clearly not integers.

Reply:
Pointers behave exactly like ptr-sized unsigned integers in casts, so it makes a lot of sense to share that code path.

Reply:
Should I investigate more or leave this alone for now? I personally didn't find this outcome to be very surprising given there are comments like this at some of the callsites:

https://github.com/rust-lang/rust/blob/f4c675c476c18b1a11041193f2f59d695b126bc8/src/librustc_mir/interpret/cast.rs#L169-L174

Reply:
I will experiment with some sanity checks in `Abi::is_signed`; I think this PR is good as-is.

Reply:
Thanks @RalfJung and @eddyb! I'll fix the tidy error tonight and `r=RalfJung,eddyb`.

Reply:
> Pointers behave exactly like ptr-sized unsigned integers in casts

Okay but why not pass `usize` to `cast_from_int` instead of the pointer type? To me the name is a lie. Maybe it could named `cast_from_int_like`.

Reply:
> Okay but why not pass usize to cast_from_int instead of the pointer type? 

I'd rather not lie about the source type.

> `cast_from_int_like`

That seems reasonable. @wesleywiser do you want to do that in this PR?

Reply:
Yeah, I can do that.